plugins/modules/foreman_scap_policy.py

```diff
@@ -0,0 +1,249 @@
+            if profile['title'] == self.params[res_type[:-1] + '_profile']:
+                return profile['id']
+        self.fail_json(msg="Can not find {0} profile ({1}) "
+                           "for the the given ({2}) {0}.".format(looking_for, self.params[res_type[:-1] + '_profile'],
+                                                                 self.params[res_type[:-1]]))
+
+
+def main():
+    module = ForemanScapPolicyModule(
+        entity_name='policy',
+        argument_spec=dict(
+            updated_name=dict(type='str'),
+        ),
+        foreman_spec=dict(
+            name=dict(type='str', required=True),
+            description=dict(type='str'),
+            deploy_by=dict(choices=['puppet', 'ansible', 'manual']),
+            scap_content=dict(type='entity', search_by='title'),
+            scap_content_profile=dict(flat_name='scap_content_profile_id', type='str'),
+            tailoring_file=dict(type='entity'),
+            tailoring_file_profile=dict(flat_name='tailoring_file_profile_id', type='str'),
+            period=dict(type='str', choices=['weekly', 'monthly', 'custom']),
+            weekday=dict(type='str', choices=[day.lower() for day in list(day_name)]),
+            day_of_month=dict(type='str', choices=[str(i) for i in range(1, 32)]),
+            cron_line=dict(type='str'),
+            hostgroups=dict(type='entity_list'),
+            hosts=dict(type='entity_list')
+        ),
+        required_if=[
+            ['period', 'weekly', ['weekday']],
+            ['period', 'monthly', ['day_of_month']],
+            ['period', 'custom', ['cron_line']],
+        ],
+        required_together=[
+            ['scap_content', 'scap_content_profile'],
+            ['tailoring_file', 'tailoring_file_profile'],
+        ],
+        required_plugins=[
+            ('openscap', ['*']),
+        ],
+    )
+
+    with module.api_connection():
+        if not module.desired_absent:
+            required_parameters = {'deploy_by', 'period', 'scap_content'}
+            missing_params = required_parameters.difference(module.foreman_params)
+            entity = module.lookup_entity('entity')
+            if entity is None and missing_params:
+                module.fail_json(msg="The following parameters are needed "
+                                     "while creating a new SCAP policy: {0}".format(", ".join(missing_params)))
+            if 'scap_content_profile' in module.foreman_params:
+                module.foreman_params['scap_content_profile'] = module.ensure_profile('title', 'scap_contents',
+                                                                                      'SCAP content')
+            if 'tailoring_file' in module.foreman_params:
+                module.foreman_params['tailoring_file_profile'] = module.ensure_profile('name', 'tailoring_files',
+                                                                                        'Tailoring file')
```

Comment:
I wonder if this is really necessary or if @mdellweg's new `lookup_entity` with a few hints in the spec is not sufficient.

Reply:
I don't see a reason, why it should not. Anyway whateverthat method is doing to find entities, it should not be named 'ensure_sth', but rather 'find_sth'.
And in case you really desperately need to find the entities by hand, you must use `module.set_entity` to put them in the `foreman_params` _and_ mark em as already resolved.

Reply:
Btw, Matthias, this new requirement is not documented anywhere ;)

Reply:
Let me fix that.

Reply:
@ogajduse how about this?

```diff
diff --git plugins/modules/foreman_scap_policy.py plugins/modules/foreman_scap_policy.py
index 53794c9..77a530e 100644
--- plugins/modules/foreman_scap_policy.py
+++ plugins/modules/foreman_scap_policy.py
@@ -182,15 +182,7 @@ from ansible.module_utils.foreman_helper import ForemanTaxonomicEntityAnsibleMod
 
 
 class ForemanScapPolicyModule(ForemanTaxonomicEntityAnsibleModule):
-    def ensure_profile(self, search_by, res_type, looking_for):
-        method_to_call = getattr(self, 'find_resource_by_' + search_by)
-        resource = method_to_call(res_type, self.params[res_type[:-1]])
-        for profile in resource[res_type[:-1] + '_profiles']:
-            if profile['title'] == self.params[res_type[:-1] + '_profile']:
-                return profile['id']
-        self.fail_json(msg="Can not find {0} profile ({1}) "
-                           "for the the given ({2}) {0}.".format(looking_for, self.params[res_type[:-1] + '_profile'],
-                                                                 self.params[res_type[:-1]]))
+    pass
 
 
 def main():
@@ -203,9 +195,9 @@ def main():
             name=dict(type='str', required=True),
             description=dict(type='str'),
             deploy_by=dict(choices=['puppet', 'ansible', 'manual']),
-            scap_content=dict(type='entity', search_by='title'),
+            scap_content=dict(type='entity', search_by='title', thin=False),
             scap_content_profile=dict(flat_name='scap_content_profile_id', type='str'),
-            tailoring_file=dict(type='entity'),
+            tailoring_file=dict(type='entity', thin=False),
             tailoring_file_profile=dict(flat_name='tailoring_file_profile_id', type='str'),
             period=dict(type='str', choices=['weekly', 'monthly', 'custom']),
             weekday=dict(type='str', choices=[day.lower() for day in list(day_name)]),
@@ -237,11 +229,24 @@ def main():
                 module.fail_json(msg="The following parameters are needed "
                                      "while creating a new SCAP policy: {0}".format(", ".join(missing_params)))
             if 'scap_content_profile' in module.foreman_params:
-                module.foreman_params['scap_content_profile'] = module.ensure_profile('title', 'scap_contents',
-                                                                                      'SCAP content')
+                scap_content = module.lookup_entity('scap_content')
+                for profile in scap_content['scap_content_profiles']:
+                    if profile['title'] == module.foreman_params['scap_content_profile']:
+                        module.foreman_params['scap_content_profile'] = profile['id']
+                        break
+                else:
+                    module.fail_json(msg="Cannot find SCAP content profile {0} for {1}".format(module.foreman_params['scap_content_profile'],
+                                                                                               module.foreman_params['scap_content']))
             if 'tailoring_file' in module.foreman_params:
-                module.foreman_params['tailoring_file_profile'] = module.ensure_profile('name', 'tailoring_files',
-                                                                                        'Tailoring file')
+                tailoring_file = module.lookup_entity('tailoring_file')
+                for profile in tailoring_file['tailoring_file_profiles']:
+                    if profile['title'] == module.foreman_params['tailoring_file_profile']:
+                        module.foreman_params['tailoring_file_profile'] = profile['id']
+                        break
+                else:
+                    module.fail_json(msg="Cannot find SCAP tailoring file profile {0} for {1}".format(module.foreman_params['tailoring_file_profile'],
+                                                                                                      module.foreman_params['tailoring_file']))
+
         module.run()
 
 
```

Reply:
Also, thanks for having failure mode tests, that saved me from posting a bad patch :)